controllers/gitrepository_controller.go

```diff
@@ -126,88 +131,139 @@ func (r *GitRepositoryReconciler) SetupWithManager(mgr ctrl.Manager) error {
 }
 
 func (r *GitRepositoryReconciler) sync(repository sourcev1.GitRepository) (sourcev1.SourceCondition, string, error) {
+	// set defaults: master branch, no tags fetching, max two commits
+	branch := "master"
+	tagMode := git.NoTags
+	depth := 2
+
 	// determine ref
-	refName := plumbing.NewBranchReferenceName("master")
-	if repository.Spec.Branch != "" {
-		refName = plumbing.NewBranchReferenceName(repository.Spec.Branch)
+	refName := plumbing.NewBranchReferenceName(branch)
+	if repository.Spec.Reference != nil {
+		if repository.Spec.Reference.Branch != "" {
+			branch = repository.Spec.Reference.Branch
+			refName = plumbing.NewBranchReferenceName(branch)
+		}
+		if repository.Spec.Reference.Commit != "" {
+			depth = 0
+		} else {
+			if repository.Spec.Reference.Tag != "" {
+				refName = plumbing.NewTagReferenceName(repository.Spec.Reference.Tag)
+			}
+			if repository.Spec.Reference.SemVer != "" {
+				tagMode = git.AllTags
+			}
+		}
+	}
+
+	// create tmp dir for SSH known_hosts
+	tmpSSH, err := ioutil.TempDir("", repository.Name)
+	if err != nil {
+		err = fmt.Errorf("tmp dir error %w", err)
+		return NotReadyCondition(sourcev1.StorageOperationFailedReason, err.Error()), "", err
 	}
-	if repository.Spec.Tag != "" {
-		refName = plumbing.NewTagReferenceName(repository.Spec.Tag)
+	defer os.RemoveAll(tmpSSH)
+
+	auth, err := r.auth(repository, tmpSSH)
+	if err != nil {
+		err = fmt.Errorf("auth error %w", err)
```

Comment:
```suggestion
		err = fmt.Errorf("auth error: %w", err)
```

The wrapping format for `fmt.Errorf` is: `<string>: %w`. See: https://blog.golang.org/go1.13-errors